### ISSUE TITLE
Show artifact type when pushing files

### ISSUE THREAD

**Issue by qweeah** [oras-project/oras#1061]
### What is the version of your ORAS CLI

1.0.1

### What would you like to be added?

When #1009 completes, by default `oras push` will build an image manifest with `artifactType` set to `--artifact-type`. If the flag is not provided, the artifact type will be default `application/vnd.unknown.artifact.v1`. 

When using `--image-spec v1.0`, the artifact type will be shown in the status of config blob `44136fa355b3`
```console
$ oras push localhost:5000/test:push foo -v --disable-path-validation --image-spec v1.0
Preparing foo
Uploading fcde2b2edba5 foo
Uploading 44136fa355b3 application/vnd.unknown.artifact.v1
Uploaded  fcde2b2edba5 foo
Uploaded  44136fa355b3 application/vnd.unknown.artifact.v1
Uploading 89773cad25ff application/vnd.oci.image.manifest.v1+json
Uploaded  89773cad25ff application/vnd.oci.image.manifest.v1+json
Pushed [registry] localhost:5000/test:push
Digest: sha256:89773cad25ffb86bf3c5a535e772ca2bd6a9f81389f598666ca70caceac5a790
```

When `--image-spec` is not provided or set to `v1.1`, the output of `oras push` doesn't mentioned the used artifact type.
```console
$ oras push localhost:5000/test:push foo -v --disable-path-validation --image-spec v1.1
Preparing foo
Uploading fcde2b2edba5 foo
Uploading 44136fa355b3 application/vnd.oci.empty.v1+json
Uploaded  fcde2b2edba5 foo
Uploaded  44136fa355b3 application/vnd.oci.empty.v1+json
Uploading d7216194f529 application/vnd.oci.image.manifest.v1+json
Uploaded  d7216194f529 application/vnd.oci.image.manifest.v1+json
Pushed [registry] localhost:5000/test:push
Digest: sha256:d7216194f52907f710f123086ea4b1bc91e0664146ac64b205560680ba4cc6bd
```

When `--image-spec` is not provided or set to `v1.1`, the output of `oras push` should contain which artifact type is used to pack the manifest, below are some suggested changes.

```console
$ # side by side with manifest media type
$ oras push localhost:5000/test:push foo -v --disable-path-validation
Preparing foo
Uploading fcde2b2edba5 foo
Uploading 44136fa355b3 application/vnd.oci.empty.v1+json
Uploaded  fcde2b2edba5 foo
Uploaded  44136fa355b3 application/vnd.oci.empty.v1+json
Uploading d7216194f529 application/vnd.oci.image.manifest.v1+json(application/vnd.unknown.artifact.v1)
Uploaded  d7216194f529 application/vnd.oci.image.manifest.v1+json(application/vnd.unknown.artifact.v1)
Pushed [registry] localhost:5000/test:push
Digest: sha256:d7216194f52907f710f123086ea4b1bc91e0664146ac64b205560680ba4cc6bd
```

Or

```console
$ # in the pushing summary below digest of manifest
$ oras push localhost:5000/test:push foo -v --disable-path-validation
Preparing foo
Uploading fcde2b2edba5 foo
Uploading 44136fa355b3 application/vnd.oci.empty.v1+json
Uploaded  fcde2b2edba5 foo
Uploaded  44136fa355b3 application/vnd.oci.empty.v1+json
Uploading d7216194f529 application/vnd.oci.image.manifest.v1+json
Uploaded  d7216194f529 application/vnd.oci.image.manifest.v1+json
Pushed [registry] localhost:5000/test:push
Digest: sha256:d7216194f52907f710f123086ea4b1bc91e0664146ac64b205560680ba4cc6bd
Artifact Type: application/vnd.unknown.artifact.v1
```

Or 
```console
$ # print when packing the manifest
$ oras push localhost:5000/test:push foo -v --disable-path-validation
Preparing foo
Preparing d7216194f529(application/vnd.unknown.artifact.v1)
Uploading fcde2b2edba5 foo
Uploading 44136fa355b3 application/vnd.oci.empty.v1+json
Uploaded  fcde2b2edba5 foo
Uploaded  44136fa355b3 application/vnd.oci.empty.v1+json
Uploading d7216194f529 application/vnd.oci.image.manifest.v1+json
Uploaded  d7216194f529 application/vnd.oci.image.manifest.v1+json
Pushed [registry] localhost:5000/test:push
Digest: sha256:d7216194f52907f710f123086ea4b1bc91e0664146ac64b205560680ba4cc6bd
```

### Are you willing to submit PRs to contribute to this feature?

- [ ] Yes, I am willing to implement it.

**Comment by FeynmanZhou** [comment]
For case 3, should we keep the consistent output format with case 1 and put the artifact type behind the digest? For example, `d7216194f529 application/vnd.unknown.artifact.v1`.

**Comment by qweeah** [comment]
> For case 3, should we keep the consistent output format with case 1 and put the artifact type behind the digest? For example, `d7216194f529 application/vnd.unknown.artifact.v1`.

Well, original `Preparing xxx` means file `xxx` is added to file store. `d7216194f529 application/vnd.unknown.artifact.v1` is a bad example since following status shows `Uploading d7216194f529 application/vnd.oci.image.manifest.v1+json`, user will wonder what's the expected media type of `d7216194f529`, is it `application/vnd.unknown.artifact.v1` or `application/vnd.oci.image.manifest.v1+json`?

Well, such question proves that proposal 3 is a bad idea.

**Comment by wangxiaoxuan273** [comment]
I'd like to work on this

**Comment by FeynmanZhou** [comment]
How about only show `artifactType` in the output when pushing an OCI v1.1 compliant artifact to the registry as follows:

- Build and push an OCI v1.1 compliant artifact, `artifactType` showed in the output:

```
oras push localhost:5000/test:push foo -v --disable-path-validation --image-spec v1.1
✓ Uploaded  foo                                                                                               91/91  B 100.00%  163ms
  └─ sha256:ab8662bef009e2ec2c84621f9e1a6564188f32bcbbebc9463c09f3113c2a9994
✓ Uploaded  application/vnd.oci.empty.v1+json                                                          2/2  B 100.00%  166ms
  └─ sha256:44136fa355b3678a1146ad16f7e8649e94fb4fc21fe77e8310c060f61caaff8a
✓ Uploaded  application/vnd.oci.image.manifest.v1+json                                                      666/666  B 100.00%     0s
  └─ sha256:5c901fd0002285d0c97efc98cfdfb1c85cc78ef93dbc45c50504462178e89a79
Pushed [registry] localhost:5000/test:push
ArtifactType: application/vnd.unknown.artifact.v1
Digest: sha256:5c901fd0002285d0c97efc98cfdfb1c85cc78ef93dbc45c50504462178e89a79
```

- Build and push an OCI v1.0 compliant artifact, no `artifactType` showed in the output:

```
oras push localhost:5000/test:push foo -v --disable-path-validation --image-spec v1.0
✓ Uploaded  foo                                                                                               91/91  B 100.00%  163ms
  └─ sha256:ab8662bef009e2ec2c84621f9e1a6564188f32bcbbebc9463c09f3113c2a9994
✓ Uploaded  application/vnd.unknown.config.v1+json                                                              2/2  B 100.00%  166ms
  └─ sha256:44136fa355b3678a1146ad16f7e8649e94fb4fc21fe77e8310c060f61caaff8a
✓ Uploaded  application/vnd.oci.image.manifest.v1+json                                                      666/666  B 100.00%     0s
  └─ sha256:5c901fd0002285d0c97efc98cfdfb1c85cc78ef93dbc45c50504462178e89a79
Pushed [registry] localhost:5000/test:push
Digest: sha256:5c901fd0002285d0c97efc98cfdfb1c85cc78ef93dbc45c50504462178e89a79
```

**Comment by FeynmanZhou** [comment]
A follow-up question, should we also display which type of content (layer, config, artifact, etc) has been uploaded in the output?

For example:

```
oras push localhost:5000/test:push foo -v --disable-path-validation --image-spec v1.1
✓ Uploaded layer foo                                                                                               91/91  B 100.00%  163ms
  └─ sha256:ab8662bef009e2ec2c84621f9e1a6564188f32bcbbebc9463c09f3113c2a9994
✓ Uploaded config application/vnd.oci.empty.v1+json                                                          2/2  B 100.00%  166ms
  └─ sha256:44136fa355b3678a1146ad16f7e8649e94fb4fc21fe77e8310c060f61caaff8a
✓ Uploaded artifact  application/vnd.oci.image.manifest.v1+json                                                      666/666  B 100.00%     0s
  └─ sha256:5c901fd0002285d0c97efc98cfdfb1c85cc78ef93dbc45c50504462178e89a79
Pushed [registry] localhost:5000/test:push
ArtifactType: application/vnd.unknown.artifact.v1
Digest: sha256:5c901fd0002285d0c97efc98cfdfb1c85cc78ef93dbc45c50504462178e89a79
```

**Comment by qweeah** [comment]
> A follow-up question, should we also display which type of content (layer, config, artifact, etc) has been uploaded in the output?

This is over designing. Also CLI don't have context of which type of content is during copying.

**Comment by qweeah** [comment]
> How about only show `artifactType` in the output when pushing an OCI v1.1 compliant artifact to the registry as follows:
> 
> * Build and push an OCI v1.1 compliant artifact, `artifactType` showed in the output:
> 
> ```
> oras push localhost:5000/test:push foo -v --disable-path-validation --image-spec v1.1
> ✓ Uploaded  foo                                                                                               91/91  B 100.00%  163ms
>   └─ sha256:ab8662bef009e2ec2c84621f9e1a6564188f32bcbbebc9463c09f3113c2a9994
> ✓ Uploaded  application/vnd.oci.empty.v1+json                                                          2/2  B 100.00%  166ms
>   └─ sha256:44136fa355b3678a1146ad16f7e8649e94fb4fc21fe77e8310c060f61caaff8a
> ✓ Uploaded  application/vnd.oci.image.manifest.v1+json                                                      666/666  B 100.00%     0s
>   └─ sha256:5c901fd0002285d0c97efc98cfdfb1c85cc78ef93dbc45c50504462178e89a79
> Pushed [registry] localhost:5000/test:push
> ArtifactType: application/vnd.unknown.artifact.v1
> Digest: sha256:5c901fd0002285d0c97efc98cfdfb1c85cc78ef93dbc45c50504462178e89a79
> ```
> 
> * Build and push an OCI v1.0 compliant artifact, no `artifactType` showed in the output:
> 
> ```
> oras push localhost:5000/test:push foo -v --disable-path-validation --image-spec v1.0
> ✓ Uploaded  foo                                                                                               91/91  B 100.00%  163ms
>   └─ sha256:ab8662bef009e2ec2c84621f9e1a6564188f32bcbbebc9463c09f3113c2a9994
> ✓ Uploaded  application/vnd.unknown.config.v1+json                                                              2/2  B 100.00%  166ms
>   └─ sha256:44136fa355b3678a1146ad16f7e8649e94fb4fc21fe77e8310c060f61caaff8a
> ✓ Uploaded  application/vnd.oci.image.manifest.v1+json                                                      666/666  B 100.00%     0s
>   └─ sha256:5c901fd0002285d0c97efc98cfdfb1c85cc78ef93dbc45c50504462178e89a79
> Pushed [registry] localhost:5000/test:push
> Digest: sha256:5c901fd0002285d0c97efc98cfdfb1c85cc78ef93dbc45c50504462178e89a79
> ```

This looks good and echos with the `ArtifactType` field in the formatted JSON output:
```console
➜  output oras version
Version:        1.2.0-beta.1
Go version:     go1.21.6
Git commit:     9ffdb3eec60b969d842af1a9e699202e0827fa01
Git tree state: clean
➜  output oras push localhost:5000/artifact:output image.tar --format json
✓ Exists    application/vnd.oci.empty.v1+json                                                      2/2  B 100.00%     0s
  └─ sha256:44136fa355b3678a1146ad16f7e8649e94fb4fc21fe77e8310c060f61caaff8a
✓ Exists    image.tar                                                                              9/9  B 100.00%     0s
  └─ sha256:8b138a688d47b23d7cc0436b95807a3f74468e8f9ed81974acc0b1b18aa6eb52
✓ Uploaded  application/vnd.oci.image.manifest.v1+json                                         587/587  B 100.00%   23ms
  └─ sha256:ff2d22f6d636cec8d11fd1a5ae15606e43989006f479ce7bf3dc78fc26e733f7
{
  "Ref": "localhost:5000/artifact@sha256:ff2d22f6d636cec8d11fd1a5ae15606e43989006f479ce7bf3dc78fc26e733f7",
  "MediaType": "application/vnd.oci.image.manifest.v1+json",
  "Digest": "sha256:ff2d22f6d636cec8d11fd1a5ae15606e43989006f479ce7bf3dc78fc26e733f7",
  "Size": 587,
  "Annotations": {
    "org.opencontainers.image.created": "2024-03-27T02:12:30Z"
  },
  "ArtifactType": "application/vnd.unknown.artifact.v1"
}
```

**Comment by qweeah** [comment]
> 
> * Build and push an OCI v1.0 compliant artifact, no `artifactType` showed in the output:
> 
> ```
> oras push localhost:5000/test:push foo -v --disable-path-validation --image-spec v1.0
> ✓ Uploaded  foo                                                                                               91/91  B 100.00%  163ms
>   └─ sha256:ab8662bef009e2ec2c84621f9e1a6564188f32bcbbebc9463c09f3113c2a9994
> ✓ Uploaded  application/vnd.unknown.config.v1+json                                                              2/2  B 100.00%  166ms
>   └─ sha256:44136fa355b3678a1146ad16f7e8649e94fb4fc21fe77e8310c060f61caaff8a
> ✓ Uploaded  application/vnd.oci.image.manifest.v1+json                                                      666/666  B 100.00%     0s
>   └─ sha256:5c901fd0002285d0c97efc98cfdfb1c85cc78ef93dbc45c50504462178e89a79
> Pushed [registry] localhost:5000/test:push
> Digest: sha256:5c901fd0002285d0c97efc98cfdfb1c85cc78ef93dbc45c50504462178e89a79
> ```

We still need to output the artifact type for v1.0 artifact since `--format json` will output artifact type
```console
➜  output oras push localhost:5000/artifact:output image.tar --format json --image-spec v1.0
✓ Exists    image.tar                                                                              9/9  B 100.00%     0s
  └─ sha256:8b138a688d47b23d7cc0436b95807a3f74468e8f9ed81974acc0b1b18aa6eb52
✓ Exists    application/vnd.unknown.config.v1+json                                                 2/2  B 100.00%     0s
  └─ sha256:44136fa355b3678a1146ad16f7e8649e94fb4fc21fe77e8310c060f61caaff8a
✓ Uploaded  application/vnd.oci.image.manifest.v1+json                                         525/525  B 100.00%     0s
  └─ sha256:ea9bd15737d671d50ed929983a73b67524464e4d96481c6d5f9c2bc261a212f3
{
  "Ref": "localhost:5000/artifact@sha256:ea9bd15737d671d50ed929983a73b67524464e4d96481c6d5f9c2bc261a212f3",
  "MediaType": "application/vnd.oci.image.manifest.v1+json",
  "Digest": "sha256:ea9bd15737d671d50ed929983a73b67524464e4d96481c6d5f9c2bc261a212f3",
  "Size": 525,
  "Annotations": {
    "org.opencontainers.image.created": "2024-03-27T02:13:42Z"
  },
  "ArtifactType": "application/vnd.unknown.config.v1+json"
}
```

Or we need to remove the `ArtifactType` from JSON output.

**Comment by qweeah** [comment]
Also, what if config data is provided and `ArtifactType` is not presented in the packed descriptor?

```console
$ oras push localhost:5000/artifact:output image.tar --format json --config <(echo -n {}):my/config
✓ Exists    my/config                                                                              0/0  B 100.00%     0s
  └─ sha256:44136fa355b3678a1146ad16f7e8649e94fb4fc21fe77e8310c060f61caaff8a
✓ Exists    image.tar                                                                              9/9  B 100.00%     0s
  └─ sha256:8b138a688d47b23d7cc0436b95807a3f74468e8f9ed81974acc0b1b18aa6eb52
✓ Uploaded  application/vnd.oci.image.manifest.v1+json                                         496/496  B 100.00%   51ms
  └─ sha256:319a41145c336132ab399512081ba7e76bb95dde261bfb399839ee2b3129b9cb
{
  "Ref": "localhost:5000/artifact@sha256:319a41145c336132ab399512081ba7e76bb95dde261bfb399839ee2b3129b9cb",
  "MediaType": "application/vnd.oci.image.manifest.v1+json",
  "Digest": "sha256:319a41145c336132ab399512081ba7e76bb95dde261bfb399839ee2b3129b9cb",
  "Size": 496,
  "Annotations": {
    "org.opencontainers.image.created": "2024-03-27T02:29:20Z"
  },
  "ArtifactType": ""
}
```

Should we keep this empty `ArtifactType` or remove it if it's not presented?

**Comment by wangxiaoxuan273** [comment]
> loaded  foo                                                                                               91/91  B 100.00%  163ms
>   └─ sha256:ab8662bef009e2ec2c84621f9e1a6564188f32bcbbebc9463c09f3113c2a9994
> ✓ Uploaded  application/vnd.unknown.config.v1+json                                                              2/2  B 100.00%  166ms
>   └─ sha256:44136fa355b3678a1146ad16f7e8649e94fb4fc21fe77e8310c060f61caaff8a
> ✓ Uploaded  applicat

I think keeping it empty is a better idea.

**Comment by sajayantony** [comment]
I'm leaning towards it omit if empty. Given that the spec is already in 1.1 and we want to focus on the 1.1 behavior going forward it would be good to just focus on ArtifactType and not auto promote configMediaType to artifactType. If we want to we could show the ConfigMediaType as well. But I don't think its that useful post 1.1 but for legacy artifact it will be.
So if we wanted to it would better as another field(which I'm still not sure is necessary)

**Comment by wangxiaoxuan273** [comment]
(editing in progress)

Final implementation and behaviors:

**Current display:**
Show ArtifactType in metadata output during `oras push`, with both `--image-spec v1.0` and `v1.1`. This is done in #1320 
![image](https://github.com/oras-project/oras/assets/103478229/09ee4bff-aa24-4581-9993-33a0849d203f)

**spec and artifact type behaviors**
This is addressed in #1317 

#1317 introduces `auto` option for `--image-spec` flag, and `auto` is made the default value. **The `auto` option makes the spec v1.1, unless --config is used and --artifact-type is not given**.

**Comment by FeynmanZhou** [comment]
After revisiting the design, `--image-spec auto` seems a bit complicated and redundant to users. In fact, the behavior of `oras push` without `--image-spec` is equal to with `--image-spec v1.1` unless only `--config` is used. Note that before [OCI image specification v1.1](https://github.com/opencontainers/image-spec/blob/v1.1.0-rc4/spec.md), the config.mediaType was used to indicate the type of the artifact (i.e. `ArtifactType`). 

I had a discussion with @wangxiaoxuan273 @qweeah , we should be able to remove `auto` from `--image-spec` and explicitly mark the default behavior of `oras push` be compliant with OCI Spec v1.1 except when `--config` is given.
